### PR TITLE
Fix Connection update over-writing Connection settings.

### PR DESCRIPTION
**Pull request by joshcanhelp**
### Changes

This fixes a bug that can cause unintended changes to database connection settings when saving a new password policy. Specific changes:

- Add object property checking to `WP_Auth0_Api_Client::update_connection()` to avoid unnecessary payload errors from the server.
- Remove duplicate connection cleanup from `WP_Auth0_Api_Operations::update_wordpress_connection()`
- Return old password policy setting if an API call fails during validation
- Pass entire connection object to the update API call to avoid unintended settings changes in password policy validation and new connection creation
- Add a large number of tests to confirm behavior
- Docblocks to help investigation and refactoring in the future

### References

- [WP forum report](https://wordpress.org/support/topic/db-migration-automaticly-disable-it-self-after-plugin-save-in-wp-admin/)

### Testing

* [x] This change adds unit test coverage
* [x] I included manual testing steps above, if applicable
* [x] This change has been tested on the latest version of the platform/language or why not

#### Scenario 1: Updates to the password policy in wp-admin

1. Start with a working WP-Auth0 install with a DB connection active
2. In the Auth0 dashboard, make sure the connection has a non-default setting (username required turned ON or custom DB migration)
3. In wp-admin, change the password policy and save

**Expected behavior:**

Password policy is changed, other settings are not affected.

#### Scenario 2: Connection creation during the setup wizard

1. Start with a WP install without WP-Auth0 installed
2. Activate the plugin and walk through the setup wizard (any option)

**Expected behavior:**

Install completes without errors in WordPress; New client created associated to new DB connection in Auth0.

### Checklist

* [x] All existing and new tests complete without errors
* [x] All code quality tools/guidelines in the [Contribution guide](CONTRIBUTION.md) have been run/followed
* [x] All active GitHub CI checks have passed

### CODE REVIEW

File: lib/WP_Auth0_Api_Operations.php [hunk 1]
```diff
@@ -30,10 +30,6 @@ public function update_wordpress_connection( $app_token, $connection_id, $passwo
 		$connection->options->customScripts->login    = $login_script;
 		$connection->options->customScripts->get_user = $get_user_script;
 
-		unset( $connection->name );
```
Comment (joshcanhelp) [inline lib/WP_Auth0_Api_Operations.php]: Moved to `WP_Auth0_Api_Client::update_connection()`

File: lib/WP_Auth0_Api_Operations.php [hunk 2]
```diff
@@ -107,26 +103,21 @@ public function create_wordpress_connection( $app_token, $migration_enabled, $pa
 		$connections   = WP_Auth0_Api_Client::search_connection( $domain, $app_token, 'auth0' );
 		$db_connection = null;
 
-		$migration_connection_id = $response->id;
+		$created_connection_id = $response->id;
```
Comment (joshcanhelp) [inline lib/WP_Auth0_Api_Operations.php]: Confusing var name

File: lib/admin/WP_Auth0_Admin_Advanced.php [hunk 3]
```diff
@@ -741,10 +741,6 @@ public function migration_ws_validation( $old_options, $input ) {
 					$connection->options->enabledDatabaseCustomization = false;
 					$connection->options->import_mode                  = false;
 
-					unset( $connection->name );
```
Comment (joshcanhelp) [inline lib/admin/WP_Auth0_Admin_Advanced.php]: Moved to `WP_Auth0_Api_Client::update_connection()`

File: lib/admin/WP_Auth0_Admin_Features.php [hunk 4]
```diff
@@ -427,8 +427,8 @@ public function security_validation( $old_options, $input ) {
 
 			foreach ( $connections as $connection ) {
 				if ( in_array( $input['client_id'], $connection->enabled_clients ) ) {
-					$patch       = array( 'options' => array( 'passwordPolicy' => $input['password_policy'] ) );
-					$update_resp = WP_Auth0_Api_Client::update_connection( $domain, $app_token, $connection->id, $patch );
+					$connection['options']['passwordPolicy'] = $input['password_policy'];
```
Comment (joshcanhelp) [inline lib/admin/WP_Auth0_Admin_Features.php]: Here's the bug ... `$patch` did not include existing Connection settings

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php [hunk 5]
```diff
@@ -128,19 +144,21 @@ public function consent_callback( $name ) {
 		$connection_exists     = false;
 		$connection_pwd_policy = null;
 
-		$connections = WP_Auth0_Api_Client::search_connection( $domain, $app_token );
+		$connections = WP_Auth0_Api_Client::search_connection( $domain, $app_token, 'auth0' );
```
Comment (joshcanhelp) [inline lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: Added strategy filter here to reduce the number of Connections returned.

File: lib/WP_Auth0_Api_Client.php [hunk 6]
```diff
@@ -619,6 +630,14 @@ public static function update_connection( $domain, $app_token, $id, $payload ) {
 		$headers['Authorization'] = "Bearer $app_token";
 		$headers['content-type']  = 'application/json';
 
+		unset( $payload->name );
```
Comment (joshcanhelp) [inline lib/WP_Auth0_Api_Client.php]: These must not be set or the API endpoint returns a payload error.

File: lib/WP_Auth0_Api_Client.php [hunk 7]
```diff
@@ -619,6 +630,14 @@ public static function update_connection( $domain, $app_token, $id, $payload ) {
...
+		unset( $payload->strategy );
+		unset( $payload->id );
+
+		if ( ! empty( $payload->enabled_clients ) ) {
```
Comment (joshcanhelp) [inline lib/WP_Auth0_Api_Client.php]: If the array is not consecutive then `json_encode` converts this to an object rather than an array.

File: lib/WP_Auth0_Api_Operations.php [hunk 8]
```diff
@@ -99,34 +95,10 @@ public function create_wordpress_connection( $app_token, $migration_enabled, $pa
...
 			return false;
-
-		}
-
-		$connections   = WP_Auth0_Api_Client::search_connection( $domain, $app_token, 'auth0' );
-		$db_connection = null;
-
-		$migration_connection_id = $response->id;
-
-		foreach ( $connections as $connection ) {
```
Comment (joshcanhelp) [inline lib/WP_Auth0_Api_Operations.php]: This method is only called in one place and the calling method already does this check.

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php [hunk 9]
```diff
@@ -128,19 +144,21 @@ public function consent_callback( $name ) {
...
 
-		foreach ( $connections as $connection ) {
+		if ( $should_create_and_update_connection && ! empty( $connections ) && is_array( $connections ) ) {
```
Comment (joshcanhelp) [inline lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: Logic here is still not great but did not want to completely refactor this in a patch release. Main thing here is checking whether we have an array and avoiding the loop if we're not using the values later.

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php [hunk 10]
```diff
@@ -128,19 +144,21 @@ public function consent_callback( $name ) {
...
 
-			if ( in_array( $client_id, $connection->enabled_clients ) ) {
-				if ( $connection->strategy === 'auth0' && $should_create_and_update_connection ) {
+				if ( in_array( $client_id, $connection->enabled_clients ) ) {
 
 					if ( $db_connection_name === $connection->name ) {
-						$connection_exists     = $connection->id;
-						$connection_pwd_policy = ( isset( $connection->options ) && isset( $connection->options->passwordPolicy ) ) ? $connection->options->passwordPolicy : null;
+						$connection_exists = $connection->id;
+						if ( isset( $connection->options ) && isset( $connection->options->passwordPolicy ) ) {
```
Comment (joshcanhelp) [inline lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: Ternary was too long and confusing to read.

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php [hunk 11]
```diff
@@ -128,19 +144,21 @@ public function consent_callback( $name ) {
...
+							$connection_pwd_policy = $connection->options->passwordPolicy;
+						}
 					} else {
-						$enabled_clients = array_diff( $connection->enabled_clients, array( $client_id ) );
-						WP_Auth0_Api_Client::update_connection( $domain, $app_token, $connection->id, array( 'enabled_clients' => array_values( $enabled_clients ) ) );
+						$connection->enabled_clients = array_diff( $connection->enabled_clients, array( $client_id ) );
```
Comment (joshcanhelp) [inline lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: Avoiding the same settings override as reported in the original bug.

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php [hunk 12]
```diff
@@ -128,19 +144,22 @@ public function consent_callback( $name ) {
...
+			foreach ( $connections as $connection ) {
 
-			if ( in_array( $client_id, $connection->enabled_clients ) ) {
-				if ( $connection->strategy === 'auth0' && $should_create_and_update_connection ) {
+				if ( in_array( $client_id, $connection->enabled_clients ) ) {
 
 					if ( $db_connection_name === $connection->name ) {
-						$connection_exists     = $connection->id;
-						$connection_pwd_policy = ( isset( $connection->options ) && isset( $connection->options->passwordPolicy ) ) ? $connection->options->passwordPolicy : null;
+						$connection_exists = $connection->id;
```
Comment (lbalmaceda) [inline lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: Can this be renamed? I get why you'd call it like that, but down in the code you store "connection exists" as the "connection id". I can't think now of a better name 😛

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php [hunk 13]
```diff
@@ -128,19 +144,22 @@ public function consent_callback( $name ) {
...
+						if ( isset( $connection->options ) && isset( $connection->options->passwordPolicy ) ) {
+							$connection_pwd_policy = $connection->options->passwordPolicy;
+						}
 					} else {
-						$enabled_clients = array_diff( $connection->enabled_clients, array( $client_id ) );
-						WP_Auth0_Api_Client::update_connection( $domain, $app_token, $connection->id, array( 'enabled_clients' => array_values( $enabled_clients ) ) );
+						$u_connection                  = clone $connection;
```
Comment (lbalmaceda) [inline lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: Add a comment above this line: "//disable connection for this application/client id"

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php [hunk 14]
```diff
@@ -191,6 +210,6 @@ public function consent_callback( $name ) {
 		}
 
 		wp_redirect( admin_url( 'admin.php?page=wpa0-setup&step=2&profile=' . $this->state ) );
-		exit();
+		exit;
```
Comment (lbalmaceda) [inline lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: was this change intended?

File: lib/admin/WP_Auth0_Admin_Features.php [hunk 15]
```diff
@@ -423,19 +423,24 @@ public function security_validation( $old_options, $input ) {
 					__( 'No database connections found for this application. ', 'wp-auth0' ) .
 					$this->get_dashboard_link( 'connections/database', __( 'See all database connections', 'wp-auth0' ) )
 				);
+				$input['password_policy'] = $old_options['password_policy'];
+				return $input;
```
Comment (joshcanhelp) [inline lib/admin/WP_Auth0_Admin_Features.php]: Exit here to skip anything else that might process (`foreach` will throw an error if it's given anything other than an array)

File: lib/admin/WP_Auth0_Admin_Features.php [hunk 16]
```diff
@@ -423,19 +423,24 @@ public function security_validation( $old_options, $input ) {
 					__( 'No database connections found for this application. ', 'wp-auth0' ) .
 					$this->get_dashboard_link( 'connections/database', __( 'See all database connections', 'wp-auth0' ) )
 				);
+				$input['password_policy'] = $old_options['password_policy'];
```
Comment (joshcanhelp) [inline lib/admin/WP_Auth0_Admin_Features.php]: Nothing changed at Auth0 so we should not change the setting in WordPress either.

File: lib/admin/WP_Auth0_Admin_Features.php [hunk 17]
```diff
@@ -423,19 +423,24 @@ public function security_validation( $old_options, $input ) {
...
+
+					$update_resp = WP_Auth0_Api_Client::update_connection( $domain, $app_token, $u_connection->id, $u_connection );
 
 					if ( false === $update_resp ) {
 						$this->add_validation_error(
 							__( 'There was a problem updating the password policy. ', 'wp-auth0' ) .
 							__( 'Please manually review and update the policy. ', 'wp-auth0' ) .
 							$this->get_dashboard_link( 'connections/database', __( 'See all database connections', 'wp-auth0' ) )
 						);
+						$input['password_policy'] = $old_options['password_policy'];
```
Comment (joshcanhelp) [inline lib/admin/WP_Auth0_Admin_Features.php]: Nothing changed at Auth0 so we should not change the setting in WordPress either.

File: lib/admin/WP_Auth0_Admin_Features.php [hunk 18]
```diff
@@ -423,19 +423,24 @@ public function security_validation( $old_options, $input ) {
 					__( 'No database connections found for this application. ', 'wp-auth0' ) .
```
Comment (lbalmaceda) [inline lib/admin/WP_Auth0_Admin_Features.php]: Can't leave a comment for the line 416 
```
 		if ( $old_options['password_policy'] !== $input['password_policy'] ) {
```
but seems like you could check the opposite value and return early instead of nesting everything here.

File: lib/admin/WP_Auth0_Admin_Features.php (hunk 17 above)
Comment (lbalmaceda) [inline lib/admin/WP_Auth0_Admin_Features.php]: I don't see where in the code you're saving the "new value" *before* doing the request, but this line seems to add nothing. If anything, the `update_connection` method should NOT modify options unless the request was successful (as part of the method logic). Why would it be able to modify the options if it hasn't checked the request results first?

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php [hunk 19]
```diff
@@ -128,19 +144,22 @@ public function consent_callback( $name ) {
...
+						$connection_exists = $connection->id;
+						if ( isset( $connection->options ) && isset( $connection->options->passwordPolicy ) ) {
+							$connection_pwd_policy = $connection->options->passwordPolicy;
+						}
 					} else {
-						$enabled_clients = array_diff( $connection->enabled_clients, array( $client_id ) );
-						WP_Auth0_Api_Client::update_connection( $domain, $app_token, $connection->id, array( 'enabled_clients' => array_values( $enabled_clients ) ) );
+						$u_connection                  = clone $connection;
+						$u_connection->enabled_clients = array_diff( $u_connection->enabled_clients, array( $client_id ) );
+						WP_Auth0_Api_Client::update_connection( $domain, $app_token, $u_connection->id, $u_connection );
```
Comment (lbalmaceda) [inline lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: what if any of this requests fail? How are you going to capture those errors and what do you plan to do next?

File: tests/testAdminFeatures.php [hunk 20]
```diff
@@ -0,0 +1,215 @@
...
+		$new_input = [
+			'password_policy' => 'good',
+			'domain'          => $test_domain,
+			'auth0_app_token' => $test_token,
+		];
+
+		$caught_request = [];
+		try {
+			$admin_features->security_validation( $old_input, $new_input );
+		} catch ( Exception $e ) {
```
Comment (lbalmaceda) [inline tests/testAdminFeatures.php]: why would the above call throw an exception?

File: tests/testAdminFeatures.php [hunk 21]
```diff
@@ -0,0 +1,215 @@
...
+		$new_input = [
+			'password_policy' => 'good',
+			'domain'          => 'test-wp.auth0.com',
+			'auth0_app_token' => implode( '.', [ uniqid(), uniqid(), uniqid() ] ),
+			'client_id'       => 'TEST_CLIENT_ID',
+		];
+
+		$this->http_request_type = [ 'success_get_connections', 'halt' ];
+
+		$caught_request = [ 'Nothing caught' ];
```
Comment (lbalmaceda) [inline tests/testAdminFeatures.php]: sounds odd that below you're treating this as a dictionary but here you store things as an array.

File: tests/testAdminFeatures.php [hunk 22]
```diff
@@ -0,0 +1,215 @@
...
+		try {
+			$admin_features->security_validation( $old_input, $new_input );
+		} catch ( Exception $e ) {
+			$caught_request = unserialize( $e->getMessage() );
+		}
+
+		$this->assertEquals( 'https://test-wp.auth0.com/api/v2/connections/TEST_CONN_ID', $caught_request['url'] );
```
Comment (lbalmaceda) [inline tests/testAdminFeatures.php]: where is the method getting the `TEST_CONN_ID` from?

File: tests/testAdminFeatures.php [hunk 23]
```diff
@@ -0,0 +1,215 @@
...
+		$this->assertEquals( $new_input['password_policy'], $result['password_policy'] );
+
+		global $wp_settings_errors;
+		$this->assertEmpty( $wp_settings_errors );
+	}
+
+	/**
+	 * Runs after each test method.
+	 */
+	public function tearDown() {
```
Comment (lbalmaceda) [inline tests/testAdminFeatures.php]: better if all the "pre/post tests" methods are together, usually on the top of the class.

File: tests/testApiOperations.php [hunk 24]
```diff
@@ -0,0 +1,173 @@
...
+	public static function setUpBeforeClass() {
+		parent::setUpBeforeClass();
+		self::$opts      = WP_Auth0_Options::Instance();
+		self::$error_log = new WP_Auth0_ErrorLog();
+	}
+
+	/**
+	 * Test that a basic create connection command requests properly.
+	 */
+	public function testThatCreateConnectionRequestsCorrectly() {
```
Comment (lbalmaceda) [inline tests/testApiOperations.php]: Odd naming. Consider the suffix you've used above "WithoutErrors". Applies to more tests on this file

File: tests/testApiOperations.php [hunk 25]
```diff
@@ -0,0 +1,173 @@
...
+	 * Setup for entire test class.
+	 */
+	public static function setUpBeforeClass() {
+		parent::setUpBeforeClass();
+		self::$opts      = WP_Auth0_Options::Instance();
+		self::$error_log = new WP_Auth0_ErrorLog();
+	}
+
+	/**
+	 * Test that a basic create connection command requests properly.
```
Comment (lbalmaceda) [inline tests/testApiOperations.php]: normally the test name is enought, no need to add docblocks on tests. Though now that you added them no need to delete them, just for future reference 👍

File: tests/testApiOperations.php [hunk 26]
```diff
@@ -0,0 +1,173 @@
...
+	 */
+	public function testThatCreateConnectionRequestsCorrectly() {
+		$this->startHttpHalting();
+
+		$api_ops    = new WP_Auth0_Api_Operations( self::$opts );
+		$test_token = implode( '.', [ uniqid(), uniqid(), uniqid() ] );
+
+		self::$opts->set( 'domain', 'test-wp.auth0.com' );
```
Comment (lbalmaceda) [inline tests/testApiOperations.php]: why are you changing the `opts` after using them to instantiate `api_ops` ?

File: tests/testApiOperations.php [hunk 27]
```diff
@@ -0,0 +1,173 @@
...
+		self::$opts->set( 'client_id', 'TEST_CLIENT_ID' );
+
+		$caught_http = [];
+		try {
+			$api_ops->create_wordpress_connection( $test_token, false, 'good' );
+		} catch ( Exception $e ) {
+			$caught_http = unserialize( $e->getMessage() );
```
Comment (lbalmaceda) [inline tests/testApiOperations.php]: minor, but you called this `caught_request` in other tests.

File: tests/testApiOperations.php [hunk 28]
```diff
@@ -0,0 +1,173 @@
...
+		parent::setUpBeforeClass();
+		self::$opts      = WP_Auth0_Options::Instance();
+		self::$error_log = new WP_Auth0_ErrorLog();
+	}
+
+	/**
+	 * Test that a basic create connection command requests properly.
+	 */
+	public function testThatCreateConnectionRequestsCorrectly() {
+		$this->startHttpHalting();
```
Comment (lbalmaceda) [inline tests/testApiOperations.php]: this sounds like a candidate for a "pre each test" method, even if not used in all the tests cases.

File: tests/testApiOperations.php [hunk 29]
```diff
@@ -0,0 +1,173 @@
...
+			$caught_http = unserialize( $e->getMessage() );
+		}
+
+		$this->assertEquals( 'https://test-wp2.auth0.com/api/v2/connections', $caught_http['url'] );
+		$this->assertEquals( 'Bearer ' . $test_token, $caught_http['headers']['Authorization'] );
+		$this->assertEquals( 'DB-Test-Blog', $caught_http['body']['name'] );
+		$this->assertEquals( 'auth0', $caught_http['body']['strategy'] );
+		$this->assertContains( 'TEST_CLIENT_ID_2', $caught_http['body']['enabled_clients'] );
+
+		$this->assertEquals( true, $caught_http['body']['options']['requires_username'] );
```
Comment (lbalmaceda) [inline tests/testApiOperations.php]: consider, if possible, using `assertTrue` instead

File: tests/testApiOperations.php [hunk 30]
```diff
@@ -0,0 +1,173 @@
...
+			$caught_http['body']['options']['customScripts']['login']
+		);
+
+		$this->assertContains(
+			'request.post("http://example.org/index.php?a0_action=migration-ws-get-user"',
+			$caught_http['body']['options']['customScripts']['get_user']
+		);
+
+		$this->assertContains(
+			'form:{username:email, access_token:"TEST_MIGRATION_TOKEN',
```
Comment (lbalmaceda) [inline tests/testApiOperations.php]: is this intentionally not having a `}` at the end of the string in test?

File: tests/testInitialSetupConsent.php [hunk 31]
```diff
@@ -0,0 +1,291 @@
...
+	public function testThatInvalidStateRedirectsProperly() {
+		$this->startRedirectHalting();
+
+		$setup_consent = new WP_Auth0_InitialSetup_Consent( self::$opts );
+		$test_domain   = 'test-wp.auth0.com';
+		$test_token    = implode( '.', [ uniqid(), uniqid(), uniqid() ] );
+
+		$caught_redirect = [];
+		try {
+			$setup_consent->callback_with_token( $test_domain, $test_token, 'invalid_state' );
```
Comment (lbalmaceda) [inline tests/testInitialSetupConsent.php]: what's the previous context? how can you ensure that `invalid_state` is an invalid state? normally, you'd manually set that state before this call.

File: tests/testInitialSetupConsent.php [hunk 32]
```diff
@@ -0,0 +1,291 @@
...
+	 */
+	public function testThatClientCreationIsAttempted() {
+		$this->startHttpHalting();
+
+		$setup_consent = new WP_Auth0_InitialSetup_Consent( self::$opts );
+		$test_token    = implode( '.', [ uniqid(), uniqid(), uniqid() ] );
+
+		$caught_http = [];
+		try {
+			$setup_consent->callback_with_token( 'test-wp.auth0.com', $test_token, 'social' );
```
Comment (lbalmaceda) [inline tests/testInitialSetupConsent.php]: how is this test logic different from the "test invalid state redirects"? Sounds like whatever caused the above test to fail should do the same here. What's changing? and if nothing is, why would an invalid_state scenario still make the create client request?

File: tests/testInitialSetupConsent.php [hunk 33]
```diff
@@ -0,0 +1,291 @@
...
+		$this->http_request_type = [
+			// Successful client creation.
+			'success_create_client',
+			// Get en existing connection enabled for this client.
+			'success_get_connections',
+			// Connection updated successfully.
+			'success_update_connection',
+			// Connection created successfully.
+			'success_create_connection',
+			// Client grant failed.
```
Comment (lbalmaceda) [inline tests/testInitialSetupConsent.php]: why would this happen? Seems like this should belong in a different test case than "create new connection succeeds"

File: tests/traits/httpHelpers.php [hunk 34]
```diff
@@ -62,7 +62,7 @@ public function stopHttpHalting() {
 	 * Use this at the top of tests that should test behavior for different HTTP responses.
 	 */
 	public function startHttpMocking() {
-		add_filter( 'pre_http_request', [ $this, 'httpMock' ], 1 );
+		add_filter( 'pre_http_request', [ $this, 'httpMock' ], 1, 3 );
```
Comment (lbalmaceda) [inline tests/traits/httpHelpers.php]: what changed here?

File: lib/WP_Auth0_Api_Client.php (hunk 7 above)
Comment (joshcanhelp) [reply lib/WP_Auth0_Api_Client.php]: If `$payload->enabled_clients` is anything other than an array with consecutive keys (like, if one of the keys in the middle is removed) then `json_encode()` turns it into an `object` which is rejected by the server.

File: lib/admin/WP_Auth0_Admin_Features.php (hunk 18 above)
Comment (joshcanhelp) [reply lib/admin/WP_Auth0_Admin_Features.php]: I could but minimal changes for the patch ... I don't want to go crazy refactoring.

File: lib/admin/WP_Auth0_Admin_Features.php (hunk 17 above)
Comment (joshcanhelp) [reply lib/admin/WP_Auth0_Admin_Features.php]: This method returns `$input`, which contains the new values. This reverts this specific value to the value it was before the admin change. API call failed so nothing should update.

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php [hunk 35]
```diff
@@ -16,6 +16,16 @@ public function __construct( WP_Auth0_Options $a0_options ) {
...
 
+	/**
+	 * Used by both Setup Wizard installation flows.
+	 * Called in WP_Auth0_InitialSetup_ConnectionProfile::callback() when an API token is used during install.
+	 * Called in self::callback() when returning from consent URL install.
+	 *
+	 * @param string $domain - Auth0 domain for the Application.
+	 * @param string $access_token - Management API access token.
+	 * @param string $type - Installation type, "social" (AKA standard) or "enterprise".
+	 * @param bool   $hasInternetConnection - True if the installing site be reached by Auth0, false if not.
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: Just documenting, not refactoring

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php (hunk 5 above)
Comment (joshcanhelp) [reply lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: Public static method so no, not currently. `WP_Auth0_Api_Client` is getting refactored one method at a time so it will go away soon.

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php (hunk 9 above)
Comment (joshcanhelp) [reply lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: I understand but looking for the minimum number of changes here. This is just one of many things that should be addressed in this method.

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php (hunk 12 above)
Comment (joshcanhelp) [reply lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: Not necessary for this PR

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php (hunk 10 above)
Comment (joshcanhelp) [reply lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: Lots of better ways to do this, yes :)

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php [hunk 36]
```diff
@@ -128,19 +144,22 @@ public function consent_callback( $name ) {
...
+				if ( in_array( $client_id, $connection->enabled_clients ) ) {
 
 					if ( $db_connection_name === $connection->name ) {
-						$connection_exists     = $connection->id;
-						$connection_pwd_policy = ( isset( $connection->options ) && isset( $connection->options->passwordPolicy ) ) ? $connection->options->passwordPolicy : null;
+						$connection_exists = $connection->id;
+						if ( isset( $connection->options ) && isset( $connection->options->passwordPolicy ) ) {
+							$connection_pwd_policy = $connection->options->passwordPolicy;
+						}
 					} else {
```
[the remark this answers is not in the document]
Comment (joshcanhelp) [reply lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: I'll keep it as-is for now. Agree that this would be better

File: lib/initial-setup/WP_Auth0_InitialSetup_Consent.php (hunk 19 above)
Comment (joshcanhelp) [reply lib/initial-setup/WP_Auth0_InitialSetup_Consent.php]: OK to continue it they fail. That method logs it internally.

File: tests/testAdminFeatures.php (hunk 20 above)
Comment (joshcanhelp) [reply tests/testAdminFeatures.php]: I'm forcing it to with `$this->startHttpHalting()` above

File: tests/testAdminFeatures.php (hunk 21 above)
Comment (joshcanhelp) [reply tests/testAdminFeatures.php]: Just a way to know if an exception was thrown or not

File: tests/testAdminFeatures.php (hunk 22 above)
Comment (joshcanhelp) [reply tests/testAdminFeatures.php]: Mocked API call

File: tests/testApiOperations.php (hunk 25 above)
Comment (joshcanhelp) [reply tests/testApiOperations.php]: Code sniffer beeps at me without it

File: tests/testApiOperations.php (hunk 28 above)
Comment (joshcanhelp) [reply tests/testApiOperations.php]: Can't halt and mock at the same time though ... some improvements are possible here, for sure.

File: tests/testApiOperations.php (hunk 30 above)
Comment (joshcanhelp) [reply tests/testApiOperations.php]: Yes, next character is a newline

File: tests/testInitialSetupConsent.php (hunk 31 above)
Comment (joshcanhelp) [reply tests/testInitialSetupConsent.php]: There are only two values that `callback_with_token` accept here ... not actually state but that's the name used in the method.

File: tests/testInitialSetupConsent.php (hunk 32 above)
Comment (joshcanhelp) [reply tests/testInitialSetupConsent.php]: `"social"` is one of two valid "state" values here so this proceeds to the API call

File: tests/testInitialSetupConsent.php (hunk 33 above)
Comment (joshcanhelp) [reply tests/testInitialSetupConsent.php]: Maybe the token doesn't have the right scopes or HTTP error or any number of things.

File: tests/traits/httpHelpers.php (hunk 34 above)
Comment (joshcanhelp) [reply tests/traits/httpHelpers.php]: Technically nothing but that last arg is the number of params that the hooked function receives.

File: tests/testInitialSetupConsent.php (hunk 33 above)
Comment (lbalmaceda) [reply tests/testInitialSetupConsent.php]: In that case, that could be moved to a different test other than the "happy path" of creating the connection when missing
